### ISSUE TITLE
Completion of error handling

### ISSUE THREAD

**Issue by elfring** [glouw/tinn#2]
Would you like to add more error handling for return values from functions like the following?
- [calloc](http://pubs.opengroup.org/onlinepubs/9699919799/functions/calloc.html "Memory allocation") ⇒ [xtbuild](https://github.com/glouw/tinn/blob/0228fdb32e89270fece8c1e186d1f891a4452f92/Tinn.c#L109)
- [fopen](http://pubs.opengroup.org/onlinepubs/9699919799/functions/fopen.html "Open a stream") ⇒ [xtsave](https://github.com/glouw/tinn/blob/0228fdb32e89270fece8c1e186d1f891a4452f92/Tinn.c#L128)

**Comment by glouw** [comment]
I'll take a handle for xtload() for fopen. Not too concerned with calloc or fopen for writes failing.

**Comment by elfring** [comment]
I suggest to avoid [ignorance of return values](https://cwe.mitre.org/data/definitions/252.html "Checking more return values") a bit more.
Would you like to detect every error situation as early as possible?

**Comment by glouw** [comment]
A little overkill in my opinion

**Comment by elfring** [comment]
I prefer a more complete error detection and exception handling.

**Comment by glouw** [comment]
Okay, I'll pop in ecalloc() and efopen() wrappers

**Comment by glouw** [comment]
assertations for callocs and fopens were added instead

**Comment by glouw** [comment]
actually, now that I think about it, asserts can be disabled globally. efopen and ecalloc were added instead.

**Comment by elfring** [comment]
Thanks for [your small source code improvement](https://github.com/glouw/tinn/commit/72dd9ed70a5bcb697b1261037592c391a71fa82c "Checking return values from two functions").
There are still [more function calls to reconsider](https://github.com/glouw/tinn/blob/2617d72ec523d4ffd89f0e6e6d6d80056326db8b/Tinn.c#L155 "Update candidate: xtsave()").

**Comment by glouw** [comment]
scanf you mean? I'll patch it

**Comment by elfring** [comment]
Functions like fclose() and fprintf() will need also further software development attention, won't they?

**Comment by glouw** [comment]
No, assuming that efopen() was successful, fclose() and fprintf() can be used without worry.

**Comment by elfring** [comment]
Your assumptions are questionable in this case. There are additional failure possibilities to consider.

How do you think about to improve [static source code analysis](https://en.wikipedia.org/wiki/List_of_tools_for_static_code_analysis#C.2C_C.2B.2B "Tools for source code analysis") also for your software?

**Comment by glouw** [comment]
I rely on runtime analysis with valgrind. Static analysis is typically overkill at times. The metrics supplied by gcc's -Wall and -Wextra is usually enough.

**Comment by elfring** [comment]
Would you like to extend the development methodology with any software tools?

**Comment by glouw** [comment]
Developments done for the most part as this is a practice in minimalism, If you're looking for a fully developed neural net library then checkout FANN (https://github.com/libfann/fann).
